File: tests/test_unit.py
#!/usr/bin/env python3
"""
MiniVault Unit Test — Python
Tests the crypto primitives directly (no interactive prompts).
Exit code 0 = pass, 1 = fail.
"""

import sys
import os
import hashlib

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from minivault import encrypt, decrypt, derive_key

PASSWORD = os.environ.get("TEST_PASSWORD", "test_password_123")
CONTENT  = os.environ.get("TEST_CONTENT", "DB_HOST=localhost\nAPI_KEY=secret\n")


def test_encrypt_decrypt_roundtrip():
    wire = encrypt(CONTENT, PASSWORD)
    assert wire, "encrypt returned empty"
    result = decrypt(wire, PASSWORD)
    assert result == CONTENT, f"Roundtrip mismatch.\nExpected: {CONTENT!r}\nGot:      {result!r}"
    print("  ✅ Roundtrip OK")


def test_wrong_password_fails():
    wire = encrypt(CONTENT, PASSWORD)
    try:
        decrypt(wire, "wrong_password")
        print("  ❌ Wrong password should have raised an error!")
        sys.exit(1)
    except (ValueError, Exception):
        print("  ✅ Wrong password correctly rejected")


def test_tampered_ciphertext_fails():
    import base64
    wire = encrypt(CONTENT, PASSWORD)
    raw  = bytearray(base64.b64decode(wire))
    # Flip a byte in the ciphertext region
    raw[-5] ^= 0xFF
    tampered = base64.b64encode(bytes(raw)).decode()
    try:
        decrypt(tampered, PASSWORD)
        print("  ❌ Tampered ciphertext should have failed integrity check!")
        sys.exit(1)
    except (ValueError, Exception):
        print("  ✅ Tampered ciphertext correctly rejected")


def test_output_is_base64():
    import base64
    wire = encrypt(CONTENT, PASSWORD)
    try:
        base64.b64decode(wire, validate=True)
        print("  ✅ Output is valid base64")
    except Exception:
        print("  ❌ Output is not valid base64")
        sys.exit(1)


def test_key_derivation():
    key = derive_key(PASSWORD)
    assert len(key) == 32, f"Key length should be 32, got {len(key)}"
    expected = hashlib.sha256(PASSWORD.encode()).digest()
    assert key == expected, "Key derivation mismatch"
    print("  ✅ Key derivation (SHA-256, 32 bytes) OK")


def test_unique_ciphertexts():
    wire1 = encrypt(CONTENT, PASSWORD)
    wire2 = encrypt(CONTENT, PASSWORD)
    assert wire1 != wire2, "Encrypting same data twice should produce different ciphertexts (random IV)"
    print("  ✅ Unique ciphertexts (random IV) OK")


if __name__ == "__main__":
    print("\n🧪 Python Unit Tests\n")
    test_key_derivation()
    test_encrypt_decrypt_roundtrip()
    test_wrong_password_fails()
    test_tampered_ciphertext_fails()
    test_output_is_base64()
    test_unique_ciphertexts()
    print("\n  All Python unit tests passed ✅\n")

File: tests/test_unit.js
#!/usr/bin/env node
/**
 * MiniVault Unit Test — Node.js
 * Tests crypto primitives directly (no interactive prompts).
 * Exit code 0 = pass, 1 = fail.
 */

"use strict";

const crypto = require("crypto");
const path   = require("path");

// ── Import crypto functions inline (same logic as minivault.js) ──

function deriveKey(password) {
  return crypto.createHash("sha256").update(password, "utf8").digest();
}

function encrypt(plaintext, password) {
  const key       = deriveKey(password);
  const iv        = crypto.randomBytes(16);
  const integrity = crypto.createHash("sha256").update(plaintext, "utf8").digest();
  const cipher    = crypto.createCipheriv("aes-256-cbc", key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  return Buffer.concat([iv, integrity, encrypted]).toString("base64");
}

function decrypt(wire, password) {
  const raw       = Buffer.from(wire, "base64");
  if (raw.length < 48) throw new Error("Archivo cifrado corrupto o demasiado corto.");
  const iv         = raw.slice(0, 16);
  const integrity  = raw.slice(16, 48);
  const ciphertext = raw.slice(48);
  const key        = deriveKey(password);
  let plaintext;
  try {
    const decipher = crypto.createDecipheriv("aes-256-cbc", key, iv);
    plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
  } catch {
    throw new Error("Contraseña incorrecta o archivo corrupto.");
  }
  const computed = crypto.createHash("sha256").update(plaintext, "utf8").digest();
  if (!crypto.timingSafeEqual(computed, integrity)) {
    throw new Error("Fallo de integridad: la contraseña es incorrecta o el archivo fue modificado.");
  }
  return plaintext;
}

// ── Tests ─────────────────────────────────────

const PASSWORD = process.env.TEST_PASSWORD || "test_password_123";
const CONTENT  = process.env.TEST_CONTENT  || "DB_HOST=localhost\nAPI_KEY=secret\n";

let failed = false;

function assert(condition, msg) {
  if (!condition) {
    console.error(`  ❌ FAIL: ${msg}`);
    failed = true;
  }
}

function test_key_derivation() {
  const key      = deriveKey(PASSWORD);
  const expected = crypto.createHash("sha256").update(PASSWORD, "utf8").digest();
  assert(key.length === 32, `Key length should be 32, got ${key.length}`);
  assert(key.equals(expected), "Key derivation mismatch");
  console.log("  ✅ Key derivation (SHA-256, 32 bytes) OK");
}

function test_roundtrip() {
  const wire   = encrypt(CONTENT, PASSWORD);
  assert(wire.length > 0, "encrypt returned empty");
  const result = decrypt(wire, PASSWORD);
  assert(result === CONTENT, `Roundtrip mismatch.\nExpected: ${CONTENT}\nGot: ${result}`);
  console.log("  ✅ Roundtrip OK");
}

function test_wrong_password() {
  const wire = encrypt(CONTENT, PASSWORD);
  try {
    decrypt(wire, "wrong_password");
    assert(false, "Wrong password should have thrown");
  } catch (_) {
    console.log("  ✅ Wrong password correctly rejected");
  }
}

function test_tampered_ciphertext() {
  const wire = encrypt(CONTENT, PASSWORD);
  const raw  = Buffer.from(wire, "base64");
  raw[raw.length - 5] ^= 0xFF;
  try {
    decrypt(raw.toString("base64"), PASSWORD);
    assert(false, "Tampered ciphertext should have failed integrity check");
  } catch (_) {
    console.log("  ✅ Tampered ciphertext correctly rejected");
  }
}

function test_output_is_base64() {
  const wire = encrypt(CONTENT, PASSWORD);
  try {
    Buffer.from(wire, "base64");
    console.log("  ✅ Output is valid base64");
  } catch (_) {
    assert(false, "Output is not valid base64");
  }
}

function test_unique_ciphertexts() {
  const w1 = encrypt(CONTENT, PASSWORD);
  const w2 = encrypt(CONTENT, PASSWORD);
  assert(w1 !== w2, "Encrypting same data twice should produce different ciphertexts (random IV)");
  console.log("  ✅ Unique ciphertexts (random IV) OK");
}

// ── Run ───────────────────────────────────────

console.log("\n🧪 Node.js Unit Tests\n");
test_key_derivation();
test_roundtrip();
test_wrong_password();
test_tampered_ciphertext();
test_output_is_base64();
test_unique_ciphertexts();

if (failed) {
  console.error("\n  Some Node.js unit tests FAILED ❌\n");
  process.exit(1);
} else {
  console.log("\n  All Node.js unit tests passed ✅\n");
}
